tapgarden: add FundBatch and SealBatch requests

jharveyb opened the pull request:
Closes #820

Update the minting state machines to move batch funding into a separate command that can execute before batch finalization. Add a new command to supply group anchor witnesses before batch finalization.

Remaining TODOs:

- [x] derive seedling script keys in `prepAssetSeedling`
- [x] DB migration to add seedling script key reference field
- [x] Update DB queries to write & fetch seedling script key
- [x] Update DB queries to write & fetch batch TX
- [x] derive group internal key in `prepAssetSeedling`
- [x] Update DB queries to write & fetch seedling group internal key
- [ ] Extend `tapdb/asset_minting` tests to cover new query params
- [ ] Add tests for `tapdb/asset_minting/CommitBatchTx`
- [x] Resolve `FinalizeParams` handling wrt. `fundBatch`
- [x] Extend `tapgarden/planter_test` test to cover `FundBatch`
- [x] Implement `SealBatch`
- [x] Remove batch ticker from caretaker and `tapgarden` tests

dstadulis commented:
Will be pulled out of draft and merged this sprint

jharveyb commented:
Feature work is done IMO, follow-on work can happen in a separate PR as this one has already ballooned quite a bit.

Remaining tasks are just unit test coverage AFAICT, should add some usage in an itest as well.

jharveyb commented on asset/asset.go:
```diff
@@ -695,6 +695,16 @@ type GroupKeyRequest struct {
 	NewAsset *Asset
 }
 
+type GroupGenesisTx struct {
```
Good point - renamed to `GroupVirtualTx` and virtual feels more consistent with the methods also used in this process + our other language around vPSBTs. I think pseudo would also fit.

jharveyb commented on fn/iter.go:
```diff
@@ -28,9 +28,9 @@ func ForEach[T any](items []T, f func(T)) {
 // effects). This can be used to ensure that any normal for-loop don't run into
 // bugs due to loop variable scoping.
-func ForEachMapItem[T any, K comparable](items map[K]T, f func(T)) {
+func ForEachMapItem[T any, K comparable](items map[K]T, f func(K, T)) {
```
It was unused before this PR, and is now used in one place:

https://github.com/lightninglabs/taproot-assets/blob/1ab44287e243bfebd7a4a4636d7ef1244540ff71/tapgarden/planter.go#L484

I think a caller could ignore the map keys with `_` if they want the old behavior.

jharveyb commented on tapgarden/planter.go:
```diff
@@ -769,9 +816,7 @@ func (c *ChainPlanter) finalizeBatch(params FinalizeParams) (*BatchCaretaker,
 
 	// At this point, we have a non-empty batch, so we'll first finalize it
 	// on disk. This means no further seedlings can be added to this batch.
-	ctx, cancel = c.WithCtxQuit()
```
I think this is resolved in later commits?

jharveyb commented:
One DB change is actually missing from this PR, related to group key handling. We need to also associate a `[32]byte` with a Seedling, which would later map to a `GroupKey.TapscriptRoot`. Without that, a user could not later provide a custom tapscript root for a group key, and if they generated witness using such a root they would not verify with the group virtual TXs made during `sealBatch`.

Roasbeef commented on tapgarden/interface.go:
```diff
@@ -43,6 +43,14 @@ type Planter interface {
 
+	// FundBatch attempts to provide a genesis point for the current batch,
+	// or create a new funded batch.
+	FundBatch(params FundParams) (*MintingBatch, error)
```
So with the the _user_ will select the genesis point? I thought it was just enough to do the coin selection earlier, then make it available to the user if needed so they can provide the custom witness?

jharveyb replied:
The `FundBatch` call, in this PR, just triggers coin selection, yes. The `FundParams` are the same as `FinalizeParams`. In some future PR we can expand those params to include a template PSBT with a genesis point (which would address #721)

Roasbeef commented on tapgarden/planter.go:
```diff
@@ -135,6 +135,27 @@ type FinalizeParams struct {
+// the genesis TX is constructed.
+type FundParams struct {
+	FeeRate        fn.Option[chainfee.SatPerKWeight]
+	SiblingTapTree fn.Option[asset.TapscriptTreeNodes]
```
Comment above says a genesis point, but looks like this is mainly about proving the sibling tap tree? Also thought we already had that plumbing hooked up here.

Roasbeef commented on tapgarden/planter.go:
```diff
@@ -135,6 +135,27 @@ type FinalizeParams struct {
+
+// groupSeal specifies the group witness for a seedling in a funded batch.
+type groupSeal struct {
+	GroupMember  asset.ID
```
Doesn't this also need the new asset name and the existing group key?

jharveyb replied:
At the point of providing witnesses, we can already display the asset ID for each seedling to the user, since the batch was already funded. And the info given when submitting the seedling is enough to map the seedling to the group it would be a part of (implemented in `buildGroupReqs`), so we don't need that info again.

Roasbeef commented on tapgarden/planter.go:
```diff
@@ -644,6 +667,29 @@ func (c *ChainPlanter) gardener() {
+				}
+
+			// TODO(jhb): follow-up PR: Implement SealBatch command
+			case reqTypeSealBatch:
```
Have it just return an error in this temp state?

Roasbeef commented on tapdb/asset_minting.go:
```diff
@@ -328,6 +328,37 @@ func (a *AssetMintingStore) CommitMintingBatch(ctx context.Context,
+					"point: %w", err)
+			}
+
+			return q.BindMintingBatchWithTx(ctx, BatchChainUpdate{
```
Passing thought: does this invalidate any existing logic that assumes that only a near finalized batch can have a batch tx bound to it?

jharveyb replied:
Good Q, so far my conclusion was no, as the funding happened in the first state of the caretaker and that field of the batch struct isn't modified before then. Since we don't unwind batch funding right now (#830), there's no logic around that to change.

We may want to delete genesis points from cancelled batches later on.

Roasbeef commented on tapgarden/caretaker.go:
```diff
@@ -383,64 +383,6 @@ func (b *BatchCaretaker) assetCultivator() {
-// (all outputs need to hold some BTC to not be dust), and with a dummy script.
-// We need to use a dummy script as we can't know the actual script key since
-// that's dependent on the genesis outpoint.
-func (b *BatchCaretaker) fundGenesisPsbt(
```
Can you add the rationale for the move to the commit body?

Roasbeef commented on tapgarden/planter.go:
```diff
@@ -773,18 +835,14 @@ func (c *ChainPlanter) gardener() {
-func (c *ChainPlanter) finalizeBatch(params FinalizeParams) (*BatchCaretaker,
-	error) {
-
+// If funding fails should we delete the batch?
```
stray comment

Roasbeef commented on tapgarden/planter.go:
```diff
@@ -773,18 +835,14 @@ func (c *ChainPlanter) gardener() {
-	error) {
-
+// If funding fails should we delete the batch?
+func (c *ChainPlanter) fundBatch(ctx context.Context, params FundParams) error {
```
Will this routine be called multiple times now? Otherwise why would the batch already be funded before this call?

Roasbeef commented on tapgarden/planter.go:
```diff
@@ -797,32 +855,140 @@ func (c *ChainPlanter) finalizeBatch(params FinalizeParams) (*BatchCaretaker,
+
+	// Process the finalize parameters.
+	params.FeeRate.WhenSome(func(fr chainfee.SatPerKWeight) {
+		feeRate = &fr
```
Can write alternatively as: 
```go
feeRate := params.FeeRate.UnwrapOr(nil)
```

Roasbeef commented on tapgarden/planter.go:
```diff
@@ -797,32 +855,140 @@ func (c *ChainPlanter) finalizeBatch(params FinalizeParams) (*BatchCaretaker,
+	ctx, cancel := c.WithCtxQuit()
+	defer cancel()
+
+	params.SiblingTapTree.WhenSome(func(tn asset.TapscriptTreeNodes) {
```
Can write alternatively as: 
```go
err := fn.MapOptionZ(params.SiblingTapTree, func(tn asset.TapscriptTreeNodes) error {
       return c.cfg.TreeStore.StoreTapscriptTree(ctx, tn)
})
```

jharveyb replied:
Since we're also returning the `rootHash` I think this doesn't work out.

Roasbeef commented on tapdb/sqlc/migrations/000017_seedling_script_group_keys.up.sql:
```diff
@@ -0,0 +1,12 @@
+-- For a group key, the internal key can also be tweaked to commit to a
+-- tapscript tree. Once we finalize the batch, this tweak will also be stored
+-- as part of the asset group itself.
+ALTER TABLE asset_seedlings ADD COLUMN group_tapscript_root BLOB;
```
Will this also eventually ref the tapscript trees we store on disk? 

Also wouldn't it be enough to just ref an group key here since that already have the sibling hash as a column?

jharveyb replied:
> Will this also eventually ref the tapscript trees we store on disk?

The value will, but this specific field won't. It'll be copied over to the final entry in the `asset_groups` table. I don't think we need to require a stored tapscript tree with a matching root though. If the user is generating a key-path witness, we only need the root and not the full tree. We only need the tree to validate a script-path witness.

"Make a group key with this tapscript tree root. If I want to produce a script-path witness, I'll later provide the tapscript tree for this tapscript tree root."

We haven't exposed the tapscript tree store yet, so the tapscript tree could be a field in the params for `sealBatch`.

> Also wouldn't it be enough to just ref an group key here since that already have the sibling hash as a column?

No, as this happens before group keys have been fully formed and inserted into `asset_groups` at all. Before being able to seal the batch, we need both an internal key for the group and this optional tweak; the asset ID can be computed after batch funding, which is the missing piece for deriving the final tweaked key.

We can only insert the asset group once we have the group witness, so at the end of `sealBatch`.

Roasbeef commented on tapdb/sqlc/queries/assets.sql:
```diff
@@ -108,11 +111,14 @@ WITH target_key_id AS (
-    sqlc.narg('group_genesis_id'), sqlc.narg('group_anchor_id')
+    sqlc.narg('group_genesis_id'), sqlc.narg('group_anchor_id'),
+    sqlc.narg('script_key_id'), sqlc.narg('group_internal_key_id'),
+   sqlc.narg('group_tapscript_root')
```
Needs an extra space to line up w/ the lines above?

Roasbeef commented on tapdb/asset_minting.go:
```diff
@@ -429,6 +434,71 @@ func (a *AssetMintingStore) CommitMintingBatch(ctx context.Context,
 }
 
+func insertOptionalSeedlingParams(ctx context.Context, q PendingAssetStore,
+	batchKey []byte, seedling *tapgarden.Seedling) ([4]sql.NullInt64,
```
How about returning a struct instead of an array? That was we can ref each of them by a field instead of the index which might change over time.

Roasbeef commented on tapdb/asset_minting.go:
```diff
@@ -543,6 +606,70 @@ func fetchAssetSeedlings(ctx context.Context, q PendingAssetStore,
+			}
+		}
+
+		if dbSeedling.GroupKeyRaw != nil {
```
We should update the unit tests assertions to make sure these are always set as expected.

Roasbeef commented on tapgarden/planter.go:
```diff
@@ -1133,6 +1133,40 @@ func (c *ChainPlanter) prepAssetSeedling(ctx context.Context,
 	}
 
+	if c.pendingBatch != nil {
+		if _, ok := c.pendingBatch.Seedlings[req.AssetName]; ok {
```
Thought we already had this check elsewhere?

jharveyb replied:
Yes, right now it's checked in `batch.addSeedling` - I added it here so it would happen earlier before we derive the script key for the seedling, but forgot to remove `addSeedling`.

Roasbeef commented on tapgarden/seedling.go:
```diff
@@ -95,6 +97,22 @@ type Seedling struct {
+	// ScriptKey is the tweaked Taproot key that will be used to spend the
+	// asset after minting. By default, this key is constructed with a
+	// BIP-0086 style tweak.
+	ScriptKey *asset.ScriptKey
```
IIUC, this'll always be set, so it doesn't need to be a pointer.

Roasbeef commented on tapgarden/seedling.go:
```diff
@@ -95,6 +97,22 @@ type Seedling struct {
+	// genesis point or tapscript root has been applied.
+	GroupInternalKey *keychain.KeyDescriptor
+
+	// TapscriptRoot is the root of the Tapscript tree that commits to all
```
`godoc` mismatch.

Roasbeef commented on tapgarden/seedling.go:
```diff
@@ -95,6 +97,22 @@ type Seedling struct {
+	// asset, these scripts are used to define witnesses more complex than
+	// a Schnorr signature for reissuing assets. A group key with an empty
+	// Tapscript root can only authorize reissuance with a signature.
+	GroupTapscriptRoot []byte
```
Any reason to not make this a `[32]byte`? Then we won't need the checks below.

jharveyb replied:
Good point, I don't think so.

jharveyb replied:
Thinking about it further, the reason we use `[]byte` for the tapscript root elsewhere is to not store the all-zero byte slice by default / for all cases where there is no root.

Roasbeef commented on tapdb/asset_minting.go:
```diff
@@ -1174,6 +1174,98 @@ func (a *AssetMintingStore) CommitBatchTx(ctx context.Context,
 
+// AddSeedlingGroups stores the asset groups for seedlings associated with a
+// batch.
+func (a *AssetMintingStore) AddSeedlingGroups(ctx context.Context,
```
Can't this be done when we commit the batch to disk to mark it as finalized?

jharveyb replied:
As in defer it to the end of `finalizeBatch` instead of at the end of `sealBatch`? If we don't write it at the end of `sealBatch` then we'd still need to persist the populated asset groups somewhere, like the batch state itself.

Though as-is we'd also want to delete the group if the batch gets cancelled, which isn't implemented.

Right now this data gets written at the end of the caretaker `BatchStateFrozen` in `AddSproutsToBatch`.

Roasbeef commented on tapgarden/planter.go:
```diff
@@ -477,6 +477,173 @@ func (c *ChainPlanter) fundGenesisPsbt(ctx context.Context,
+	seedlings map[string]*Seedling) (map[string]*Seedling,
+	map[string]*Seedling) {
+
+	WithGroup := make(map[string]*Seedling)
```
Variable names should start with a lower case character.

Roasbeef commented on tapgarden/planter.go:
```diff
@@ -903,7 +1070,86 @@ func (c *ChainPlanter) fundBatch(ctx context.Context, params FundParams) error {
 }
 
-func (c *ChainPlanter) sealBatch(params SealParams) error {
+func (c *ChainPlanter) sealBatch(ctx context.Context, _ SealParams) error {
```
Doesn't use the param?

jharveyb replied:
Deferred for a follow-up PR.

Roasbeef commented on tapgarden/planter.go:
```diff
@@ -903,7 +1070,86 @@ func (c *ChainPlanter) fundBatch(ctx context.Context, params FundParams) error {
+			return err
+		}
+
+		err = c.cfg.TxValidator.Execute(groupedAsset, nil, nil)
```
Where's the witness? Also comments here aren't full sentences above.

jharveyb replied:
The witness is a field of `groupKey` above, and we have to re-make the asset with the witness populated to actually validate.

jharveyb commented:
Updated to address some feedback and update tests to exercise query changes. Also fixed some issues with not accepting externally-set seedling script keys and group internal keys.

Hitting some issues with minting DB tests around using the seedling group tapscript root.

From discussion offline, I think the simplest behavior for `FundBatch` and `SealBatch` would be that they can be called exactly once before `Finalize`. Its straightforward to detect if a batch is funded or sealed, so it should also be straightforward to return an error here and require the batch to be resubmitted.

guggero commented on tapdb/asset_minting.go:
```diff
@@ -408,35 +415,38 @@ func (a *AssetMintingStore) CommitMintingBatch(ctx context.Context,
+				return err
+			}
+
+			dbSeedling.GroupInternalKeyID =
```
Much nicer with the named struct!

guggero commented on tapdb/assets_common.go:
```diff
@@ -84,21 +84,55 @@ type UpsertAssetStore interface {
 	genesisPoint, err := encodeOutpoint(genesisOutpoint)
 	if err != nil {
-		return 0, fmt.Errorf("unable to encode genesis point: %w", err)
+		return 0, fmt.Errorf("%w: %w", ErrEncodeGenesisPoint, err)
```
I know newer Golang versions allow for double wrapping errors. But not sure how that works exactly? Do we really want to do that?

jharveyb replied:
Good catch, I tested this in the playground and what I added doesn't work as expected. I mostly wanted to dedupe these error messages; AFAICT callers right now aren't inspecting the concrete error type.

So I think the fix would be to just define these new variables as strings instead of errors, and use:

`fmt.Errorf("%s: %w)`.

Callers would then have to use `strings.Contains()` instead of `errors.Is()`.

guggero commented on asset/asset.go:
```diff
@@ -695,6 +695,16 @@ type GroupKeyRequest struct {
 	NewAsset *Asset
 }
 
+type GroupVirtualTx struct {
```
nit: Godoc missing (struct and members).

guggero commented on asset/asset.go:
```diff
@@ -1045,14 +1055,18 @@ func (req *GroupKeyRequest) Validate() error {
-func DeriveGroupKey(genSigner GenesisSigner, genBuilder GenesisTxBuilder,
-	req GroupKeyRequest) (*GroupKey, error) {
-
+// Build the virtual TX needed to produce an asset group witness.
```
nit: Godoc comment doesn't match function name.

guggero reviewed (approved):
LGTM, pending existing comments and successful CI run.

jharveyb commented:
Added logic to reject `FundBatch` and `FinalizeBatch` calls when the pending batch is already funded.
Updated minting DB tests for the new seedling key handling, and for better coverage.

We had some behavior around group verifiers that assumed that the group tapscript root was always empty, which is fixed.
I also removed the old static `DeriveGroupKey` func, as it is no longer appropriate with all the new seedling options (and it wasn't being tested in the VM tests either).

Also added a test around making a batch with external seedling keys and actually using `FundBatch`.

Moving on to the follow-up PR but will some spots here like new queries around seedling groups still need tests added.

jharveyb commented:
Rebased.

Roasbeef reviewed (approved):
LGTM 🦚

guggero commented:
Needs another rebase to apply new linter rules, then a couple of fixes for the new rules.
